metric/system/cgroup/reader.go

```diff
@@ -0,0 +1,365 @@
+	subsystems, err := SupportedSubsystems(opts.RootfsMountpoint)
+	// We can return a not-quite-an-error ErrCgroupsMissing here, so return the bare error.
+	if err != nil {
+		return nil, err
+	}
+
+	// Locate the mountpoints of those subsystems.
+	mountpoints, err := SubsystemMountpoints(opts.RootfsMountpoint, subsystems)
+	if err != nil {
+		return nil, fmt.Errorf("error finding mountpoints: %w", err)
+	}
+
+	return &Reader{
+		rootfsMountpoint:         opts.RootfsMountpoint,
+		ignoreRootCgroups:        opts.IgnoreRootCgroups,
+		cgroupsHierarchyOverride: opts.CgroupsHierarchyOverride,
+		cgroupMountpoints:        mountpoints,
+	}, nil
+}
+
+// CgroupsVersion reports if the given PID is attached to a V1 or V2 controller
+func (r *Reader) CgroupsVersion(pid int) (CgroupsVersion, error) {
+	cgPath := filepath.Join("/proc/", strconv.Itoa(pid), "cgroup")
+	cgPath = r.rootfsMountpoint.ResolveHostFS(cgPath)
+	cgraw, err := ioutil.ReadFile(cgPath)
+	if err != nil {
+		return CgroupsV1, fmt.Errorf("error reading %s: %w", cgPath, err)
+	}
+	cgstring := string(cgraw)
+	//V2 cgroups always begin with 0::/
+	// Some distros will "mix" V1 and V2 by adding an unused V2 controller
+	// Check to see if we're actually using V2.
+	// This is cautious code, as different distros seem to mix V2 and V1 events in weird ways.
+	if strings.Contains(cgstring, "0::/") {
+		// This is a V2-only cgroup
+		if len(strings.Split(strings.TrimSpace(cgstring), "\n")) == 1 {
+			return CgroupsV2, nil
+		}
+		// Otherwise, check to see what's in the controllers file
+		controllers, err := readControllerList(cgstring, r.cgroupMountpoints.V2Loc)
+		if err != nil {
+			return CgroupsV1, fmt.Errorf("error fetching cgroup controller list for pid %d: %w", pid, err)
+		}
+		// The logic here is a tad opinionated. If we're at this point in the code, it's because we have both
+		// V1 and V2 controllers on a cgroup. If the V2 controller has no actual controllers associated with it,
+		// We revert to V1. If it does, report V2. In the future, we may want to "combine" V2 and V1 metrics somehow.
+		if len(controllers) > 0 {
+			logp.L().Debugf("fetching V2 controller: %#v for pid %d\n", controllers, pid)
+			return CgroupsV2, nil
+		}
+		return CgroupsV1, nil
+	}
+	return CgroupsV1, nil
+}
+
+// GetStatsForPid is a generic method that returns a CGStats interface for V1 and V2
+// cgroup statistics. For applications that require raw metrics, use GetV*StatsForProcess()
+func (r *Reader) GetStatsForPid(pid int) (CGStats, error) {
+	v, err := r.CgroupsVersion(pid)
+	if err != nil {
+		return nil, fmt.Errorf("error finding cgroup version for pid %d: %w", pid, err)
+	}
+	if v == CgroupsV1 {
+		return r.GetV1StatsForProcess(pid)
+	}
+	return r.GetV2StatsForProcess(pid)
+}
+
+// GetV1StatsForProcess returns cgroup metrics and limits associated with a process.
+func (r *Reader) GetV1StatsForProcess(pid int) (*StatsV1, error) { // nolint: dupl // return value is different
```

Comment:
This `dupl` warning is odd. This particular linter seems to have a high false positive rate.